tfs/src/main/java/hudson/plugins/tfs/listeners/JenkinsRunListener.java

```diff
@@ -54,4 +88,46 @@ private String getStartedBy(final Run run) {
         }
         return startedBy;
     }
+
+    private TeamGitStatus trySetPullRequestStatus(final Run run, @Nonnull final TaskListener listener, final GitStatusState buildState, final String buildDescription, final String targetUrl) {
+        try {
+            final TeamGitStatus status = new TeamGitStatus();
+            status.state = buildState;
+            status.description = buildDescription;
+            status.targetUrl = targetUrl;
+            status.context = new GitStatusContext("ci-build", "jenkins-plugin");
+
+            if (!UnsupportedIntegrationAction.isSupported(run, listener)) {
+                final PrintStream logger = listener.getLogger();
+                logger.print("NOTICE: ");
+                logger.print("You selected '");
+                logger.print(targetUrl);
+                logger.println("' on your Jenkins job, but this option has no effect when calling the job from the 'Jenkins Queue Job' task in TFS/Team Services.");
+                return null;
+            }
+
+            final CommitParameterAction commitParameter = run.getAction(CommitParameterAction.class);
+            final GitCodePushedEventArgs gitCodePushedEventArgs;
+            final PullRequestMergeCommitCreatedEventArgs pullRequestMergeCommitCreatedEventArgs;
+            if (commitParameter != null) {
+                gitCodePushedEventArgs = commitParameter.getGitCodePushedEventArgs();
+                if (commitParameter instanceof PullRequestParameterAction) {
+                    final PullRequestParameterAction prpa = (PullRequestParameterAction) commitParameter;
+                    pullRequestMergeCommitCreatedEventArgs = prpa.getPullRequestMergeCommitCreatedEventArgs();
+                    final URI collectionUri = gitCodePushedEventArgs.collectionUri;
+                    final TeamRestClient client = new TeamRestClient(collectionUri);
+                    return client.addPullRequestStatus(pullRequestMergeCommitCreatedEventArgs, status);
+                }
+            }
+        } catch (MalformedURLException e) {
+            e.printStackTrace();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+    private JSONObject createJsonFromRun(final Run run) {
+        return new JSONObject();
```

Comment:
This just returns an empty json object.  is this correct?

Reply:
That's what I can do for now as it's undecided which fields are required in VSTS. If Jason finalizes the fields that should be passed to VSTS, either of us should completely remove getApiJson API, and include all required fields in this created json object.

Reply:
Is it worth checking this is in until then though?  This feature won't be usable in any meaningful installation until then, I think.

Reply:
Actually two different events are sent in OnComplete:

1. JobCompletionEvent event, which posts a Jenkins build result json to VSTS (https://github.com/smile21prc/tfs-plugin/blob/master/tfs/src/main/java/hudson/plugins/tfs/util/TeamRestClient.java#L311). This step is not part of my feature. It's triggered for every Jenkins build, regardless whether it's a PR CI build or non-PR build. That one won't work perfectly as expected until the json file is properly constructed.

2. PullRequestBuildStatusEvent event -- this is my feature, which posts PR build status back to corresponding VSTS PR API, ONLY for those jobs triggered by a PR.  As it's independent from the above event, it works regardless whether the above event is correctly sent. So the check-in of my feature shouldn't be blocked by the decision on proper handling of the above event.